src/math_util.c

```diff
@@ -166,6 +167,48 @@ dot_qd(const qd *A, const qd *B, qd *C) {
     c_qd_add(tmp[3], tmp[2], C->x);
 }
 
+/*
+    normalized_dot_qd returns dot product of normalized input vectors.
+*/
+static NPY_INLINE int
+normalized_dot_qd(const qd *A, const qd *B, qd *dot_val) {
+    int i, flag;
+    qd aa, bb, ab;
+    double aabb[4];
+    double norm[4];
+    double *v0 = dot_val->x;
+    double *v1 = dot_val->x + 1;
+    double eps = 10.0 * c_qd_epsilon();
+
+    dot_qd(A, A, &aa);
+    dot_qd(B, B, &bb);
+    dot_qd(A, B, &ab);
+    c_qd_mul(aa.x, bb.x, aabb);
+
+    if (aabb[0] < -0.0) {
```

Comment:
I don't understand this part where the product of the dot products of each vector is computed. In principle each dot product should be positive (or zero), but never negative, and yet there is a test for a negative value. If there is a way for the dot product to be negative through a bug, then both might be negative. And what does this have to do with normalizing with regard to the dot of A and B? I am missing something.

Reply:
> And what does this have to do with normalizing with regard to the dot of A and B? I am missing something.

It has nothing to do with normalizing with regard to the dot of A and B. Yes, you are missing the comment above the function:

> normalized_dot_qd returns dot product of **normalized input vectors**.

The original code was doing this:
```
Anormalized = A / SQRT(A.A)
Bnormalized = B / SQRT(B.B)
AB = Anormalized.Bnormalized
```

Modified code does this:
```
AB = A.B / SQRT((A.A)*(B.B))
```

Reply:
So I was missing something.